### PR TITLE
Add trait groups.

### PR DESCRIPTION
**Pull request by SpencerMichaels**
This PR is split off from #22487. It depends on #22495 (hopefully I did the latter part right).

Previously, when specifying an NPC class's traits, one could only do so through a format like this:
```
{
    "type" : "npc_class",
    ... more fields here ...
    "traits" : [
        [ "GOODHEARING", 100 ],
        [ "STRONGBACK", 60 ]
    ]
}
```
This is not very flexible, however. Consider, for instance, a case where a modder wants to make an NPC that has Genetically Unstable 2/3rds of the time, and Genetic Chaos 1/3 of the time — but never both at once. With the old system, this is not possible. With this new system, you can do this:
```
{
    "type" : "npc_class",
    ... more fields here ...
    "subtype": "distribution",
    "entries" : [
        [ "trait": "CHAOTIC", prob: 1 ],
        [ "trait": "UNSTABLE", prob: 2 ]
    ]
}
```
What if you want to give the same set of mutations to multiple NPCs, or choose between multiple sets of mutations at random? You can define a `trait_group` — whose syntax is quite similar to an `item_group` — and nest arbitrarily complicated invocations of collections and distributions. Collections will choose whether or not to add a trait based on its probability considered in insolation. Distributions will chose _one_ of their traits based on the weights of all the traits summed together. For instance, the example group below is a collection containing a trait and a distribution, with the distribution containing two traits. It will _always_ generate trait `OUTDOORSMAN`, and half of the time will generate _either_ `GOODCARDIO` (2/3rds chance) or `GOODHEARING` (1/3rd chance).
```
  {
    "type" : "trait_group",
    "id" : "outdoorsman_distribution",
    "subtype": "collection",
    "entries": [
        { "trait": "OUTDOORSMAN" },
        { "distribution" : [
            { "trait": "GOODCARDIO", "prob": 60 },
            { "trait": "GOODHEARING", "prob": 30 }
        ], "prob": 50}
    ]
  },
```

This also adds functionality to the trait blacklist, making blacklisted traits disappear from any trait groups that contain them. 

# Screenshots
This is the result of a simple roll of 100% genetically unstable and 50% genetic chaos

This is the result of the `outdoorsman_distribution` trait group example above.

This is the group automatically generated by the `traits` entry in `NC_APIS` (see `json/npcs/classes.json:394`), the bee mutant NPC class.

### CODE REVIEW

File: src/trait_group.h [hunk 1]
```diff
@@ -0,0 +1,197 @@
+#pragma once
+#ifndef TRAIT_GROUP_H
+#define TRAIT_GROUP_H
+
+#include "json.h"
+#include "mutation.h"
+#include "string.h"
+#include "string_id.h"
```
Comment (BevapDin) [inline src/trait_group.h]: You don't need to include "json.h" here, forward declarations of the used JSON classes are enough for the header. Probably the same for the "mutation.h". And we don't have a "string.h" header. You should include \<string\>, \<vector\> and \<memory\> here as their content is used here. Don't rely on other headers to include what your header uses.

**Comment by kevingranade** [comment]
Tried to merge, spawned some NPCs, didn't end up spawning with traits.  I don't know if I merged correctly.

**Comment by sxmichaels** [comment]
Er, this doesn't add NPCs with traits in the sense that they will now spawn
with traits by default. There are no new NPC classes added, it just adds
the framework necessary to give NPCs traits via class definitions. You
would need to test this via the new debug menu option I added (debug -->
test trait groups, I think), which I think will at least have one entry it
picks up from NC_APIS (the bee mutant already in the game). You could also
spawn an NPC and change its class to Apis but I think the result will look
the same as if you did it in the current master, since it's just
under-the-hood stuff that changed in this PR.

If you'd like, I could add some new NPC classes to this PR that use the new
trait system, or put up a separate PR with a mod like I did for the
mutation rounds feature.

(Sorry for the lack of responses on my PRs recently, by the way, I've been
finishing up this semester.)

2017年12月14日 00:52，"Kevin Granade" <notifications@github.com>寫道：

Tried to merge, spawned some NPCs, didn't end up spawning with traits. I
don't know if I merged correctly.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/CleverRaven/Cataclysm-DDA/pull/22496#issuecomment-351624382>,
or mute the thread

**Comment by kevingranade** [comment]
It does add new NPC classes, and I verified that some random NPCs spawned with those classes,  but they had no traits. 

> (Sorry for the lack of responses on my PRs recently, by the way, I've been finishing up this semester.)

No problem at all, interruptions are expected.

**Comment by sxmichaels** [comment]
Oh, huh, so it does… well, I could have sworn I'd removed that test mod, but apparently not. I cleaned it up a bit just now, anyway.

With regards to random NPCs not spawning with traits (good catch, by the way, it didn't even occur to me that would be any different!), it turned out I was using `type` rather than `myclass` in `npc::randomize()`. The issue was that the former is not populated with the right values in random spawns but does have what it needs in debug spawns. This PR fixes it.

**Comment by kevingranade** [comment]
Now I can't get the new NPC class to spawn at all.
Also when I run the trait group test it has odd results,
group 4 says 100x invincibility???
group 5 lists 100x each of the same 11 mutations, which are all mutant things like spiked tail or antenna, not traits.

**Comment by sxmichaels** [comment]
What do you mean by "can't get the class to spawn"? Possibly it's a merge issue — I just merged the current master into my branch and tested it again, perhaps this will work for you. I made a fresh world with a very high NPC spawn rate and the `NPC traits` mod enabled, and I did indeed run into random NPCs with the new class, and they have traits.

The test output sounds like it's working correctly, though. Those two groups you mentioned are picked up from `NC_TEST_CLASS` and `NC_APIS` respectively; the former has a 100% chance of spawning with invincibility and the latter a 100% chance of spawning with various insect mutations. What I called "traits" in this PR actually refers to both mutations and starting traits, since both were listed under the tag `traits` in the JSON under the old system.

**Comment by BorkBorkGoesTheCode** [comment]
@sxmichaels ?

**Comment by sxmichaels** [comment]
As far as I can tell the most recent comment doesn't have the issues
@kevingranade is describing (see my post above) so I was hoping he could
explain further / provide confirmation. I don't think there's anything for
me to do right now.

2018年1月15日 01:44，"BorkBorkGoesTheCode" <notifications@github.com>寫道：

@sxmichaels <https://github.com/sxmichaels> ?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/CleverRaven/Cataclysm-DDA/pull/22496#issuecomment-357604579>,
or mute the thread

**Pull request merged**